File: harness/codegen/control_flow/goblint.py
from typing import Iterable
from harness.core import Process
from harness.control_flow import ControlFlowMutex
from harness.codegen.control_flow.base import HarnessControlFlowBaseCodegen, IntOrStrIterable

class HarnessControlFlowGoblintCodegen(HarnessControlFlowBaseCodegen):
    def _prologue(self):
        yield from self._embed_multiline('''
extern _Atomic long _harness_random;

typedef unsigned int __harness_thread_t;
typedef unsigned int __harness_mutex_t;
extern void __harness_thread_create(__harness_thread_t *, void *, void *(*)(void *), void *);
extern void __harness_thread_join(__harness_thread_t, void **);
extern void __harness_mutex_init(__harness_mutex_t *, void *);
extern void __harness_mutex_lock(__harness_mutex_t *);
extern void __harness_mutex_unlock(__harness_mutex_t *);
''')

    def _declare_mutex(self, mutex: str):
        yield f'static __harness_mutex_t {mutex};'

    def _declare_init_barrier(self, processes: Iterable[Process]):
        for process in processes:
            yield f'static _Atomic unsigned int process_{process.mnemonic}_init_barrier;'

    def _open_process_definition(self, process: Process):
        yield f'void *process_{process.mnemonic}(void *arg) {{'
        yield 1
        yield '(void) arg; // Unused'

    def _close_process_definition(self, process):
        yield 'return NULL;'
        yield -1
        yield '}'

    def _open_main_definition(self):
        yield 'int main() {'
        yield 1

    def _close_main_definition(self):
        yield 'return 0;'
        yield -1
        yield '}'

    def _setup_init_barrier(self, processes: Iterable[Process]):
        for process in processes:
            yield f'process_{process.mnemonic}_init_barrier = 0;'
    
    def _init_barrier_wait(self, process: Process, other_processes: Iterable[Process]) -> IntOrStrIterable:
        yield f'process_{process.mnemonic}_init_barrier = 1;'
        yield 'while ({}); // Wait for other processes'.format(
            ' || '.join(
                f'!process_{other_process.mnemonic}_init_barrier'
                for other_process in other_processes
            )
        )

    def _initialize_mutex(self, mutex: str):
        yield f'__harness_mutex_init(&{mutex}, NULL);'

    def _declare_process(self, process: Process):
        yield f'__harness_thread_t process_thread_{process.mnemonic};'

    def _start_process(self, process: Process):
        yield f'__harness_thread_create(&process_thread_{process.mnemonic}, NULL, process_{process.mnemonic}, NULL);'

    def _join_process(self, process: Process):
        yield f'__harness_thread_join(process_thread_{process.mnemonic}, NULL);'

    def _lock_mutex(self, mutex: str):
        yield f'__harness_mutex_lock(&{mutex});'

    def _unlock_mutex(self, mutex: str):
        yield f'__harness_mutex_unlock(&{mutex});'
    
    def _mutex_set_transition(self, lock: Iterable[ControlFlowMutex], unlock: Iterable[ControlFlowMutex]) -> IntOrStrIterable:
        lock_mtx = sorted(
            (
                mtx
                for mtx in lock
            ), key=lambda mtx: mtx.identifier
        )
        unlock_mtx = reversed(sorted(
            (
                mtx
                for mtx in unlock
            ), key=lambda mtx: mtx.identifier
        ))

        for mtx in lock_mtx:
            yield from self._lock_mutex(self._mutex_name(mtx))
        for mtx in unlock_mtx:
            yield from self._unlock_mutex(self._mutex_name(mtx))

    def _random(self, max):
        return f'(_harness_random % {max})'

File: harness2/src/harness/codegen/goblint.rs
use crate::harness::{control_flow::mutex::ControlFlowMutexID, core::{error::HarnessError, process::ProcessID}};

use super::base::{CodegenOutput, ControlFlowCodegen};

pub struct ControlFlowGoblintCodegen;

impl ControlFlowGoblintCodegen {
    pub fn new() -> ControlFlowGoblintCodegen {
        ControlFlowGoblintCodegen {}
    }
}

impl<Output: CodegenOutput> ControlFlowCodegen<Output> for ControlFlowGoblintCodegen {
    fn generate_prologue(&self, output: &mut Output) -> Result<(), HarnessError> {
        self.embed_multiline(output, r#"
extern _Atomic long _harness_random;

typedef unsigned int __harness_thread_t;
typedef unsigned int __harness_mutex_t;
extern void __harness_thread_create(__harness_thread_t *, void *, void *(*)(void *), void *);
extern void __harness_thread_join(__harness_thread_t, void **);
extern void __harness_mutex_init(__harness_mutex_t *, void *);
extern void __harness_mutex_lock(__harness_mutex_t *);
extern void __harness_mutex_unlock(__harness_mutex_t *);
"#)?;
        Ok(())
    }
    
    fn declare_mutex(&self, output: &mut Output, mutex: ControlFlowMutexID) -> Result<(), HarnessError> {
        output.write_line(format!("static __harness_mutex_t mutex{};", Into::<u64>::into(mutex)))?;
        Ok(())
    }

    fn declare_init_barrier(&self, output: &mut Output, processes: impl Iterator<Item = ProcessID>) -> Result<(), HarnessError> {
        for process in processes {
            output.write_line(format!("static _Atomic unsigned int process{}_init_barrier;",
                Into::<u64>::into(process)))?;
        }
        Ok(())
    }

    fn begin_process_definition(&self, output: &mut Output, process: ProcessID) -> Result<(), HarnessError> {
        output.write_line(format!("void *process{}(void *arg) {{", Into::<u64>::into(process)))?;
        output.indent_up();
        output.write_line("(void) arg; // Unused")?;
        Ok(())
    }

    fn end_process_definition(&self, output: &mut Output, _: ProcessID) -> Result<(), HarnessError> {
        output.write_line("return NULL;")?;
        output.indent_down();
        output.write_line("}")?;
        Ok(())
    }

    fn begin_main_definition(&self, output: &mut Output) -> Result<(), HarnessError> {
        output.write_line("int main() {")?;
        output.indent_up();
        Ok(())
    }

    fn end_main_definition(&self, output: &mut Output) -> Result<(), HarnessError> {
        output.write_line("return 0;")?;
        output.indent_down();
        output.write_line("}")?;
        Ok(())
    }

    fn setup_init_barrier(&self, output: &mut Output, processes: impl Iterator<Item = ProcessID>) -> Result<(), HarnessError> {
        for process in processes {
            output.write_line(format!("process{}_init_barrier = 0;",
                Into::<u64>::into(process)))?;
        }
        Ok(())
    }

    fn initialize_mutex(&self, output: &mut Output, mutex: ControlFlowMutexID) -> Result<(), HarnessError> {
        output.write_line(format!("__harness_mutex_init(&mutex{}, NULL);",
            Into::<u64>::into(mutex)))?;
        Ok(())
    }

    fn declare_process_thread(&self, output: &mut Output, process: ProcessID) -> Result<(), HarnessError> {
        output.write_line(format!("__harness_thread_t process_thread{};",
            Into::<u64>::into(process)))?;
        Ok(())
    }

    fn start_process_thread(&self, output: &mut Output, process: ProcessID) -> Result<(), HarnessError> {
        output.write_line(format!("__harness_thread_create(&process_thread{}, NULL, process{}, NULL);",
            Into::<u64>::into(process), Into::<u64>::into(process)))?;
        Ok(())
    }

    fn join_process_thread(&self, output: &mut Output, process: ProcessID) -> Result<(), HarnessError> {
        output.write_line(format!("__harness_thread_join(process_thread{}, NULL);",
            Into::<u64>::into(process)))?;
        Ok(())
    }

    fn lock_mutex(&self, output: &mut Output, mutex: ControlFlowMutexID) -> Result<(), HarnessError> {
        output.write_line(format!("__harness_mutex_lock(&mutex{});",
            Into::<u64>::into(mutex)))?;
        Ok(())
    }

    fn unlock_mutex(&self, output: &mut Output, mutex: ControlFlowMutexID) -> Result<(), HarnessError> {
        output.write_line(format!("__harness_mutex_unlock(&mutex{});",
            Into::<u64>::into(mutex)))?;
        Ok(())
    }

    fn do_synchronization(&self, output: &mut Output, lock: impl Iterator<Item = ControlFlowMutexID>, unlock: impl Iterator<Item = ControlFlowMutexID>, _: Option<&str>) -> Result<(), HarnessError> {
        for mutex in lock {
            self.lock_mutex(output, mutex)?;
        }
        for mutex in unlock {
            self.unlock_mutex(output, mutex)?
        }
        Ok(())
    }

    fn wait_init_barrier(&self, output: &mut Output, process: ProcessID, other_processes: impl Iterator<Item = ProcessID>) -> Result<(), HarnessError> {
        output.write_line(format!("process{}_init_barrier = 1;",
            Into::<u64>::into(process)))?;

        let mut condition = String::new();
        for (index, other_process) in other_processes.enumerate() {
            if index > 0 {
                condition.push_str(" || ");
            }
            condition.push_str(&format!("!process{}_init_barrier",
                Into::<u64>::into(other_process)));
        }

        output.write_line(format!("while ({});", condition))?;
        Ok(())
    }

    fn generate_random(&self, max: u32) -> String {
        format!("(_harness_random % {})", max)
    }
}
